scipy/spatial/transform/rotation.py

```diff
@@ -0,0 +1,754 @@
+
+        In terms of DCMs, this application is the same as
+        `self.as_dcm().dot(vectors)`.
+
+        The number of rotations and number of vectors given must follow
+        standard numpy broadcasting rules: either one of them equals unity or
+        they both equal each other.
+
+        Returns a `numpy.ndarray` of shape `(3,)` if object contains a single
+        rotation (as opposed to a stack with a single rotation) and a single
+        vector is specified with shape `(3,)`. In all other cases, the returned
+        array has shape `(N, 3)`, where `N` is either the number of rotations
+        or vectors.
+
+        Parameters
+        ----------
+        vectors : array_like, shape (3,) or (N, 3)
+            Each `vectors[i]` represents a vector in 3D space. A single vector
+            can either be specified with shape `(3, )` or `(1, 3)`.
+        inverse : boolean, optional
+            If `inverse` is `True` then the inverse of the rotation(s) is
+            applied to the input vectors. Default is `False`.
+        """
+        vectors = np.asarray(vectors)
+        if vectors.ndim > 2 or vectors.shape[-1] != 3:
+            raise ValueError("Expected input of shape (3,) or (P, 3), "
+                             "got {}.".format(vectors.shape))
+
+        single_vector = False
+        if vectors.shape == (3,):
+            single_vector = True
+            vectors = vectors[None, :]
+
+        dcm = self.as_dcm()
+        if self._single:
+            dcm = dcm[None, :, :]
+
+        n_vectors = vectors.shape[0]
+        n_rotations = self.n
+
+        if n_vectors != 1 and n_rotations != 1 and n_vectors != n_rotations:
+            raise ValueError("Expected equal numbers of rotations and vectors "
+                             ", or a single rotation, or a single vector, got "
+                             "{} rotations and {} vectors.".format(
+                                n_rotations, n_vectors))
+
+        if inverse:
+            result = np.einsum('ikj,ik->ij', dcm, vectors)
+        else:
+            result = np.einsum('ijk,ik->ij', dcm, vectors)
+
+        if self._single and single_vector:
+            return result[0]
+        else:
+            return result
+
+    def __getitem__(self, indexer):
+        """Extract rotation at given index(es) from object.
+
+        This function returns a new `Rotation` instance containing:
+
+            - a single rotation, if `indexer` is a single index
+            - a stack of rotation(s), if `indexer` is a slice, or an index
+              array.
+
+        A single indexer must be specified, i.e. as if indexing a 1 dimensional
+        array or list.
+
+        Parameters
+        ----------
+        indexer : index, slice, or index array
+            Specifies which rotation(s) to extract.
+        """
+        # __init__ now copies by default
+        return self.__class__(self._quat[indexer], normalized=True)
```

Comment:
this should have a `.copy()` because if `indexer` makes a view (e.g., is a slice) then the new object will not own its own data, i.e.:
```
>>> x = np.zeros(3)
>>> y = np.asarray(x, float)
>>> y[2] = 1
>>> x
array([0., 0., 1.])
```
It would be good to add a tiny test case that exposes this problem, and then fix it (true TDD!)

Reply:
We've already taken care of this in the `__init__` function. Even if `self._quat[indexer]` returns a view (which it does in most cases if my understanding of python indexing and slicing is correct), the `self.__class__()` calls `__init__` where we make a copy of the input quaternions before setting the private `_quat` variable. 

This has been verified by the following test case, which I will include in the `test_rotation` file:
```python
>>> r = R.from_quaternion(np.random.uniform(size=(3,4)))
>>> r._quat
array([[0.61921279, 0.49144496, 0.592531  , 0.15480437],
       [0.47882848, 0.60457664, 0.24404027, 0.58792407],
       [0.10501252, 0.1217286 , 0.25670971, 0.95302395]])
>>> s = r[0:2]
>>> s._quat
array([[0.61921279, 0.49144496, 0.592531  , 0.15480437],
       [0.47882848, 0.60457664, 0.24404027, 0.58792407]])
>>> r._quat[0] = [0,0,0,1]
>>> r._quat
array([[0.        , 0.        , 0.        , 1.        ],
       [0.47882848, 0.60457664, 0.24404027, 0.58792407],
       [0.10501252, 0.1217286 , 0.25670971, 0.95302395]])
>>> s._quat
array([[0.61921279, 0.49144496, 0.592531  , 0.15480437],
       [0.47882848, 0.60457664, 0.24404027, 0.58792407]])
```
The new `s` object still contains its initial quaternions and is not modified when `r` is.